Source/santad/Logs/EndpointSecurity/Writers/SpoolTest.mm

```diff
@@ -0,0 +1,135 @@
+#include "Source/santad/Logs/EndpointSecurity/Writers/FSSpool/fsspool_log_batch_writer.h"
+#include "Source/santad/Logs/EndpointSecurity/Writers/Spool.h"
+
+namespace santa::santad::logs::endpoint_security::writers {
+
+class SpoolPeer : public Spool {
+ public:
+  // Make constructors visible
+  using Spool::Spool;
+
+  std::string GetTypeUrl() { return type_url_; }
+};
+
+}  // namespace santa::santad::logs::endpoint_security::writers
+
+using ::fsspool::FsSpoolLogBatchWriter;
+using ::fsspool::FsSpoolWriter;
+using santa::santad::logs::endpoint_security::writers::Spool;
+using santa::santad::logs::endpoint_security::writers::SpoolPeer;
+
+@interface SpoolTest : XCTestCase
+@property dispatch_queue_t q;
+@property dispatch_source_t timer;
+@property NSFileManager *fileMgr;
+@property NSString *testDir;
+@property NSString *baseDir;
+@property NSString *spoolDir;
+@end
+
+@implementation SpoolTest
+
+- (void)setUp {
+  self.fileMgr = [NSFileManager defaultManager];
+  self.testDir = [NSString stringWithFormat:@"%@santa-spool-%d", NSTemporaryDirectory(), getpid()];
+  self.testDir = [NSString stringWithFormat:@"%@fsspool-%d", NSTemporaryDirectory(), getpid()];
+  self.baseDir = [NSString stringWithFormat:@"%@/base", self.testDir];
+  self.spoolDir = [NSString stringWithFormat:@"%@/new", self.baseDir];
+
+  XCTAssertFalse([self.fileMgr fileExistsAtPath:self.baseDir]);
+  XCTAssertFalse([self.fileMgr fileExistsAtPath:self.spoolDir]);
+
+  XCTAssertTrue([self.fileMgr createDirectoryAtPath:self.testDir
+                        withIntermediateDirectories:YES
+                                         attributes:nil
+                                              error:nil]);
+
+  self.q = dispatch_queue_create(NULL, DISPATCH_QUEUE_SERIAL);
+  XCTAssertNotNil(self.q);
+  self.timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.q);
+  XCTAssertNotNil(self.timer);
+}
+
+- (void)tearDown {
+  XCTAssertTrue([self.fileMgr removeItemAtPath:self.testDir error:nil]);
+}
+
+- (void)testTypeUrl {
+  // Ensure the manually created type url isn't modified
+  auto spool =
+    std::make_shared<SpoolPeer>(self.q, self.timer, [self.baseDir UTF8String], 10240, 1024);
+  std::string wantTypeUrl("type.googleapis.com/santa.pb.v1.SantaMessage");
+  XCTAssertCppStringEqual(spool->GetTypeUrl(), wantTypeUrl);
+}
+
+- (void)testWrite {
+  const useconds_t waitTime = 1000000;      // 1.0 seconds
+  const useconds_t shortWaitTime = 250000;  // 0.25 seconds
+  const size_t writeSize = 50;
+  const uint64 periodicFlushMS = 400;
+  NSError *err = nil;
+
+  auto spool =
+    std::make_shared<SpoolPeer>(self.q, self.timer, [self.baseDir UTF8String], 10240, 1024);
+
+  // Set a custom timer interval for this test
+  dispatch_source_set_timer(self.timer, dispatch_time(DISPATCH_TIME_NOW, 0),
+                            NSEC_PER_MSEC * periodicFlushMS, 0);
+
+  spool->Write(std::vector<uint8_t>(writeSize, 'A'));
+
+  // Sleep for 1.25 seconds
+  usleep(waitTime);
```

Comment:
Sleeps in test code always make me uneasy. I think we can get around this using the callback we talked about in chat.